qbraid_qir/qasm3/transformer.py

```diff
@@ -0,0 +1,212 @@
+        step = (
+            1 if range_def.step is None else Qasm3ExprEvaluator.evaluate_expression(range_def.step)
+        )
+        Qasm3Validator.validate_register_index(start_qid, qreg_size, qubit=is_qubit_reg)
+        Qasm3Validator.validate_register_index(end_qid - 1, qreg_size, qubit=is_qubit_reg)
+        return list(range(start_qid, end_qid, step))
+
+    @staticmethod
+    def transform_gate_qubits(
+        gate_op: QuantumGate, qubit_map: dict[str, IndexedIdentifier]
+    ) -> None:
+        """Transform the qubits of a gate operation with a qubit map.
+
+        Args:
+            gate_op (QuantumGate): The gate operation to transform.
+            qubit_map (dict[str, IndexedIdentifier]): The qubit map to use for transformation.
+
+        Returns:
+            None
+        """
+        for i, qubit in enumerate(gate_op.qubits):
+            if isinstance(qubit, IndexedIdentifier):
+                raise_qasm3_error(
+                    f"Indexing '{qubit.name.name}' not supported in gate definition",
+                    span=qubit.span,
+                )
+
+            gate_op.qubits[i] = qubit_map[qubit.name]
+
+    @staticmethod
+    def transform_gate_params(
+        gate_op: QuantumGate, param_map: dict[str, Union[FloatLiteral, IntegerLiteral]]
+    ) -> None:
+        """Transform the parameters of a gate operation with a parameter map.
+
+        Args:
+            gate_op (QuantumGate): The gate operation to transform.
+            param_map (Dict[str, Union[FloatLiteral, IntegerLiteral]]): The parameter map to use
+                                                                        for transformation.
+
+        Returns:
+            None
+        """
+        for i, param in enumerate(gate_op.arguments):
+            if isinstance(param, Identifier):
+                gate_op.arguments[i] = param_map[param.name]
+            # TODO : update the arg value in expressions not just SINGLE identifiers
+
+    @staticmethod
+    def get_branch_params(
+        condition: Union[UnaryExpression, BinaryExpression, IndexExpression]
+    ) -> tuple[Optional[int], Optional[str]]:
+        """
+        Get the branch parameters from the branching condition
+
+        Args:
+            condition (Union[UnaryExpression, BinaryExpression, IndexExpression]): The condition
+                                                                                   to analyze
+
+        Returns:
+            tuple[int, str]: The branch parameters
+        """
+        if isinstance(condition, UnaryExpression):
+            return (
+                condition.expression.index[0].value,
+                condition.expression.collection.name,
+            )
+        if isinstance(condition, BinaryExpression):
+            return condition.lhs.index[0].value, condition.lhs.collection.name
+        if isinstance(condition, IndexExpression):
+            return condition.index[0].value, condition.collection.name
+        return None, None
```

Comment:
Based on the type hint `condition: Union[UnaryExpression, BinaryExpression, IndexExpression]`, this line is unreachable